Variance-covariance testing

nathanielolin opened the pull request:
- Adds a test for custom variance-covariance matrices and degrees-of-freedom for T statistics
- Warns if you specify `vcov` but not `dof` and your model is OLS
- Adds two small data objects for the test (plus code to make them): a vcov matrix + dof from a clustered OLS model, and a vcov matrix + dof from a logit model

#33, #55

nathanielolin commented on R/mod_marg.R:
```diff
@@ -35,9 +36,20 @@
                       type = 'levels',
-                      vcov_mat = vcov(mod),
+                      vcov_mat = NULL,
+                      dof = NULL,
```
Made this change (and lines 75-79 below) because otherwise, to test if the user specified a variance-covariance matrix, you have to call vcov(mod) again and check for equality with vcov_mat.

anniejw6 commented on R/mod_marg.R:
```diff
@@ -11,6 +11,7 @@
 #' Defaults to 1.
 #' @param at_var_interest vector, if type == 'levels', the values for the variable of interest at which levels should be calculated.
 #' If NULL, indicates all levels for a factor variable, defaults to NULL
+#' @param dof integer, the degrees of freedom used for the T statistic in an OLS model
```
Needs to say defaults to NULL and what that defaults to

anniejw6 commented on R/mod_marg.R:
```diff
@@ -35,9 +36,20 @@
 mod_marg2 <- function(mod, var_interest,
                       type = 'levels',
-                      vcov_mat = vcov(mod),
+                      vcov_mat = NULL,
```
Need to fix in documentation that it no longer defaults to `vcov(mod)`

nathanielolin replied:
Done.

anniejw6 commented on R/mod_marg.R:
```diff
@@ -52,6 +64,20 @@ mod_marg2 <- function(mod, var_interest,
 
+  if(is.null(dof) & !is.null(vcov_mat) & mod$family$family == 'gaussian')
+    warning(paste(
+      "You provided a new variance-covariance matrix for an OLS model",
```
I'm not sure whether this belongs as a warning or as a note under `Details`. I'm just not familiar enough with DOF corrections.

nathanielolin replied:
Changed to be a bit shorter and reference the details in the documentation. I think the warning is useful, because it's not something people think about, and the estimates / SE's replicate fine so it's easy to miss the incorrect P value.

anniejw6 commented on data-raw/make_cluster_vcov.R:
```diff
@@ -0,0 +1,50 @@
+
```
Need to add this folder to `.Rbuildignore`, e.g., https://github.com/hadley/babynames/blob/master/.Rbuildignore

See also http://r-pkgs.had.co.nz/data.html

anniejw6 replied:
I also think this would be cool to add to a vignette, maybe the main one, maybe a secondary one.

nathanielolin replied:
added to .Rbuildignore; good idea for the vignette (but maybe for a new MR)

anniejw6 commented on R/cvcov.R:
```diff
@@ -0,0 +1,16 @@
+#' Clustered variance-covariance matrices and T statistic d.o.f.
```
Did you want this to be exported to users?

nathanielolin replied:
I think so; it's used in an example and someone might be like "what the heck is this"

anniejw6 replied:
Okay.

anniejw6 commented on R/pred_se_wrap.R:
```diff
@@ -32,7 +33,8 @@
                          at_var_interest = NULL,
-                         vcov_mat = NULL){
+                         vcov_mat = NULL,
+                         dof = model$df.residual){
```
We're in dependency hell a little bit with these interior functions. What is the norm here for what these things should default to? Since the defaults are not the same as the default of the function that calls them. I just worry.

nathanielolin replied:
Let's change `pred_se_wrap` to not have default arguments for `vcov_mat` or `dof`, since it's not designed to be called independently by the user.

anniejw6 commented on R/mod_marg.R:
```diff
@@ -35,9 +36,20 @@
 #'            data = margex, family = 'binomial')
 #' mod_marg2(mod, var_interest = 'sex', type = 'levels', at = NULL)
+#'
+#' data(margex)
```
You should add a note here describing what this example does (i.e., custom variance-covariance matrix).

I would also add to details.

nathanielolin replied:
Done.

anniejw6 commented on data-raw/make_cluster_vcov.R:
```diff
@@ -0,0 +1,50 @@
+
+# Create clustered variance-covariance matrix for package tests
+
+library(modmarg)
```
Why?

anniejw6 commented on data-raw/make_cluster_vcov.R:
```diff
@@ -0,0 +1,50 @@
+library(modmarg)
+library(sandwich)
+library(data.table)
+
```
Can we call functions from those packages instead of loading the entire namespace?

anniejw6 commented on data-raw/make_cluster_vcov.R:
```diff
@@ -0,0 +1,50 @@
+rm(list = ls())
+
+cluster_se <- function(model, cluster, data = model$data){
+  data <- data[complete.cases(data[, names(model$model)]), ]
```
Does this work with interaction terms? (Might want to write a test.)

nathanielolin replied:
Great question. I'll check.

anniejw6 commented on data-raw/make_cluster_vcov.R:
```diff
@@ -0,0 +1,50 @@
+
+# Create clustered variance-covariance matrix for package tests
+
```
Is there a citation or something? Doesn't have to be but would be nice, I think.

anniejw6 commented on data-raw/make_cluster_vcov.R:
```diff
@@ -0,0 +1,50 @@
+}
+
+# Create sample vcov matrix + degrees of freedom for vcov test
+# Section D.1 here:
```
What is D.1? This appears to be ordered by Chapter, then section... But there is no section D in chapter 1. Page number?

anniejw6 commented on tests/testthat/test_vcov.R:
```diff
@@ -0,0 +1,42 @@
+
+  data(margex)
+  mod <- glm(outcome ~ treatment + distance, data = margex, family = 'gaussian')
+
```
Need stata stuff + stata output.

anniejw6 commented on tests/testthat/test_vcov.R:
```diff
@@ -0,0 +1,42 @@
+  # stata
+  # logit outcome i.treatment distance, vce(cluster arm)
+  # margins i.treatment"]
+
```
Can we add stata output?

anniejw6 commented on tests/testthat/test_vcov.R:
```diff
@@ -0,0 +1,42 @@
+
+test_that("clustered standard errors are correct", {
+
+  data(margex)
```
can you add a comment here that you're just comparing gaussian vs binomial for these two models? it just took me a while to figure out what's going on

anniejw6 replied:
(fixed)

anniejw6 commented on DESCRIPTION:
```diff
@@ -1,13 +1,13 @@
 Package: modmarg
 Title: Calculating Marginal Effects and Levels with Errors
-Version: 0.2.0
+Version: 0.2.1
```
Bump up the middle one - this adds functionality, not just a bug fix.

nathanielolin commented:
Addressed your comments save three:

- Need a general citation for clustered SEs
- Need to check if it works with interaction terms
- Might want to add to vignette.

I think the vignette can happen in another MR. The first I just need to find something concise; the second will take me a bit more time (might have to happen later today).

nathanielolin commented on R/output.R:
```diff
@@ -38,8 +40,8 @@ format_output <- function(margin_labels, pred_margins, se, family, dof,
-    upper_ci = pred_margins + (qnorm(max(cofint)) * se )
+    lower_ci = pred_margins + min(ci) * se,
+    upper_ci = pred_margins + max(ci) * se
   )
```
If we're using a T statistic for P values, need to do so for CIs too.

anniejw6 commented on R/pred_se_wrap.R:
```diff
@@ -0,0 +1,86 @@
+#' Calculate margins and se
```
@nathanielolin - you need to kill this file.

anniejw6 commented on R/cvcov.R:
```diff
@@ -0,0 +1,16 @@
+#' See data-raw/make_cluster_vcov.R for details.
+#'
+#' @format A list of two lists, one for OLS and one for logit, each containing
+#' \describe{
```
You need to add the documentation about the "weird" one -- and why is it weird?

anniejw6 commented on data-raw/make_cluster_vcov.R:
```diff
@@ -0,0 +1,60 @@
+
+mod <- glm(mpg ~ cyl * poly(disp, degree = 2, raw = TRUE) + hp,
+           data = mtcars)
+weird_cvcov <- cluster_se(mod, "gear")
```
why "weird"? Can we something more descriptive?

anniejw6 commented on tests/testthat/test_vcov.R:
```diff
@@ -0,0 +1,130 @@
+  expect_equal(z$Margin, c(0.0802249, 0.2588702), tolerance = 0.0001)
+  expect_equal(z$Standard.Error, c(0.0481047, 0.0467188),
+               tolerance = 0.0001)
+  expect_equal(z$P.Value, c(0.2373067, 0.0310606), tolerance = 0.0001)
```
I don't see these values in the stata output...

anniejw6 commented on tests/testthat/test_vcov.R:
```diff
@@ -0,0 +1,130 @@
+  expect_equal(z$Margin, c(0.0791146, 0.2600204), tolerance = 0.0001)
+  expect_equal(z$Standard.Error, c(0.0473008, 0.0472214),
+               tolerance = 0.0001)
+  expect_equal(z$P.Value, c(9.4409e-02, 3.6600e-08), tolerance = 0.0001)
```
Don't see this precision in the stata output

anniejw6 commented on tests/testthat/test_vcov.R:
```diff
@@ -0,0 +1,130 @@
+               tolerance = 0.0001)
+  expect_equal(z$Test.Stat, c(0.890104, 6.5334911, 25.3411205),
+               tolerance = 0.0001)
+  expect_equal(z$P.Value, c(0.4673266, 0.0226343, 0.0015536),
```
Don't see this precision in the stata output

anniejw6 commented:
@nathanielolin - still a few changes needed.

anniejw6 commented:
@nathanielolin - this LGTM. I wasn't sure if you were done, so merge whenever you're ready.
